### ISSUE TITLE
Class Counter in Sigmoid Cross Entropy Loss

### ISSUE THREAD

**Issue by ppwwyyxx** [s9xie/hed#7]
```cpp
int dim = bottom[0]->count() / bottom[0]->num();
for (int i = 0; i < num; ++i) {
      temp_loss_pos = 0;
      temp_loss_neg = 0;
      for (int j = 0; j < dim; j ++) {
         if (target[i*dim+j] == 1) {
        	count_pos ++;
        	temp_loss_pos -= input_data[i*dim + j] * (target[i*dim+j] - (input_data[i*dim + j] >= 0)) -
                	log(1 + exp(input_data[i*dim + j] - 2 * input_data[i*dim + j] * (input_data[i*dim + j] >= 0)));
    	}
    	else if (target[i*dim+j] == 0) {
        	count_neg ++;
        	temp_loss_neg -= input_data[i*dim + j] * (target[i*dim+j] - (input_data[i*dim + j] >= 0)) -
                	log(1 + exp(input_data[i*dim + j] - 2 * input_data[i*dim + j] * (input_data[i*dim + j] >= 0)));
    	}
     } 
     loss_pos += temp_loss_pos * count_neg / (count_pos + count_neg);
     loss_neg += temp_loss_neg * count_pos / (count_pos + count_neg);
}
```
The above code in [this file](https://github.com/s9xie/hed/blob/master/src/caffe/layers/sigmoid_cross_entropy_loss_layer.cpp#L50) seems confusing. Although I guess this won't make a huge difference.

In every iteration the counter `count_pos` and `count_neg` are accumulated. These two counters are not set to zero in each iteration, but they are used in each iteration. This doesn't look right.

Also, from the paper it looks like the ratio should be a constant that is calculated across all data in the training set.

**Comment by s9xie** [comment]
The accumulation iters through all spatial locations in one SINGLE image. The counter resets everytime the forward function is called. The balancing factor we used is not precomputed across the whole dataset (for which it could be), but on a single image instead.

**Comment by ppwwyyxx** [comment]
I was not clear. By "iteration", I mean the iteration of the outer for-loop in the code I posted.
Since the counters were not reset between these iterations, each image in a batch ends up using a different `beta`, which is neither the neg/pos ratio within a batch, nor a neg/pos ratio within an image.

**Comment by s9xie** [comment]
Oooops. Thanks for the heads-up. You are correct. This is a bug. Luckily it doesn't affect the results: because it is in the forward loop only for "showing" the loss. And the (wrongly) computed balancing factor, should still be around 0.05vs0.95, similar to what is computed globally, or on a single image.   

The crucial backward computation is correct. 

Thanks again for noticing this... I'll correct it right away.

**Comment by lexie0212** [comment]
quick question on the backward computation: on line 107:
bottom_diff[i * dim + j] *= 1 * count_neg / (count_pos + count_neg);

and line 110
bottom_diff[i * dim + j] *= count_pos / (count_pos + count_neg);

why there is an extra "1" in line 107? It looked like doesn't make any differences between these two lines while only bottom_diff is weighted based on the count of labels?

**Comment by ppwwyyxx** [comment]
Yes. You can remove the "1" and it's obviously equivalent.

**Comment by brisker** [comment]
@ppwwyyxx 
Hi, the modified SigmoidCrossEntropy Loss layer has a [line](https://github.com/s9xie/hed/blob/master/src/caffe/layers/sigmoid_cross_entropy_loss_layer.cpp#L107)
that reads like -:

    bottom_diff[i * dim + j] *= 1 * count_neg / (count_pos + count_neg);

why not 

     bottom_diff[i * dim + j] = 1 * count_neg / (count_pos + count_neg);
?
It seems that the gradients are multiplied every iteration in the loops. It is a little confusing to me.  Why does the author write like that?

**Comment by ppwwyyxx** [comment]
That's back propagation.

**Comment by brisker** [comment]
@ppwwyyxx 
The loss may be something like:
![image](https://cloud.githubusercontent.com/assets/13804492/24184098/3a889f3a-0f07-11e7-98a8-772599c7bfa4.png)
So you mean the gradients calculation is automatically done by caffe, here in the code , the  author is just multiplying the gradients by the weight factor?

**Comment by ppwwyyxx** [comment]
It's done in the code. https://github.com/s9xie/hed/blob/master/src/caffe/layers/sigmoid_cross_entropy_loss_layer.cpp#L89
